Stabilization v2

sxzhang25 opened the pull request:
Add v2 stabilization using the isStabilized bit.

sxzhang25 committed:
Check do_not_stabilize bit for v2.

sxzhang25 committed:
Refactor BufferProcessor into MotionPhotoReaderUtils.java

The reason for this refactor is that the BufferProcessor class shared many variables with MotionPhotoReader, so it was easier to write code if the process() function in BufferProcessor (which handled calls to nextFrame() and seekTo() synchronously) was moved back into the respective methods in MotionPhotoReader. The remaining methods that BufferProcessor contains also were not all necessarily related to processing the buffer queues, so it made more sense to make it a general reader utilities class.

- Move process() in BufferProcessor back into nextFrame() and seekTo().
- Move all shared variables back into MotionPhotoReader and pass them into the methods of MotionPhotoReaderUtils as arguments instead.
- Make all methods in MotionPhotoReaderUtils public static.
- Update MotionPhotoReaderTest to reflect the refactor.
- Create new tests under MotionPhotoUtilsTest (previously BufferProcessorTest).

sxzhang25 committed:
Merge branch 'stabilization-v1' into refactor-reader-utils

sxzhang25 committed:
Update unit and instrumentation tests for MotionPhotoReader and MotionPhotoReaderUtils.

- Add new constants for MotionPhotoReaderUtilsTest.
- Fix constructors and open() in MotionPhotoReaderTest.

sxzhang25 committed:
Clean up code.

- Remove unused constructors and completed TODOs in HomographyMatrix.
- Clarify Javadoc in MotionPhotoInfo.

sxzhang25 committed:
Clean up code and clarify Javadoc.

- Remove unused constructors and completed TODOs in HomographyMatrix.
- Fix stabilization on/off logic in seekTo() and nextFrame().

sxzhang25 committed:
Fix seekTo() track selection.

sxzhang25 committed:
Add rotation matrix creator in HomographyMatrix.

sxzhang25 committed:
Add scale matrix creator in HomographyMatrix.

sxzhang25 committed:
Merge branch 'stabilization-v1' into stabilization-v2

sxzhang25 committed:
Fix v2 stabilization (skip over variable size block in 0xC5, otherwise isStabilized is always set to true).

sxzhang25 committed:
Add v2 stabilization.

sxzhang25 committed:
Merge remote-tracking branch 'origin' into refactor-reader-utils

# Conflicts:
#	MotionPhotoReader/app/src/androidTest/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderTest.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/Constants.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/HomographyMatrix.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java
#	MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoWidget.java
#	MotionPhotoReader/app/src/test/java/com/google/android/libraries/motionphotoreader/BufferProcessorTest.java

sxzhang25 committed:
Fix code style.

sxzhang25 committed:
Fix code style and update getHomographies() to cover the case when the protocol buffer is unavailable. Specifically, create a variable which stores the last set of homography data extracted successfully from a protocol buffer and use this as a default.

sxzhang25 committed:
Fix homography matrix tests.

Note: Arrays.asList(...) does not make a copy of the array passed.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java:
```diff
@@ -26,11 +25,13 @@
+ *   height: The height of the video (before camera orientation rotations are applied) in pixels.
+ *   durationUs: The duration of the video in microseconds.
+ *   rotation: The camera orientation while the motion photo was taken, as a rotation about the
+ *   z-axis, measured in degrees.
```
Is that a cw or ccw rotation?

sxzhang25 replied:
Clockwise, where 0 degrees corresponds to holding the phone horizontally (top of the phone on the left). I added in that note.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+    public static Integer getInputBuffer(BlockingQueue<Integer> availableInputBuffers) {
+        Integer bufferIndex = -1;
+        try {
+            bufferIndex = availableInputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
```
There appears to be a time unit inconsistency here (µs vs ms).

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+        try {
+            bufferIndex = availableInputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
+        } catch (InterruptedException e) {
+            Log.e(TAG, "No input buffer available", e);
```
In this code path, we return `-1` which is undocumented by Javadoc. Should we also return the special value `null` in that case?

sxzhang25 replied:
Right, I think it should be null (made a change to initialize bufferIndex as null).

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+    public static Bundle getOutputBuffer(BlockingQueue<Bundle> availableOutputBuffers) {
+        Bundle bufferData = null;
+        try {
+            bufferData = availableOutputBuffers.poll(TIMEOUT_US, TimeUnit.MILLISECONDS);
```
Same time unit inconsistency as above.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+            // If the protocol buffer was not null, then extract the homography data list from the
+            // protocol buffer data; otherwise, just use the homography data list from the most
+            // recent non-null protocol buffer
+            List<Float> homographyDataList = (protocolBufferIsAvailable) ?
```
Parentheses can be removed.

clju commented on MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+                    protocolBufferIsAvailable = false;
+                }
+            } catch (InvalidProtocolBufferException e) {
+                Log.e(TAG, "Could not parse from protocol buffer");
```
`protocolBufferIsAvailable` should be set to false too in that case too. Or better yet, it should be initialized to false and be only explicitly set to true in the code paths where you know it's valid.

sxzhang25 committed:
Minor fixes:

- Fix code style.
- Update logic for protocolBufferIsAvailable in getHomographies().
- Rename TIMEOUT_US to TIMEOUT_MS for time unit consistency.
